### DISCUSSION TITLE
Setting up a node with gprc on a distributed server

### DISCUSSION THREAD

**Discussion by awolff64** [coreemu/core#757] [Q&A] [answered]
Config:
- main machine: current core daemon and core gui on Ubuntu 22.04 in VM on a Ubuntu 22.10 host.
- remote: current core daemon  in linux container with Ubuntu 22.04 on a Ubuntu 22.10 host.

Configuring nodes via the GUI works fine (after a fix, see [here](https://github.com/coreemu/core/discussions/756)).

I now try to set up a configuration using grpc:

1. I run the python script (modified example) below on my main machine.
2. This should create on node on the main and one on remote.
3. First I got an error that the server "WAN" is not known. After some digging in the code I added the line that adds the servers to the session.
4. Now the two nodes are created - but on the main machine. I can see this by connecting with the GUI to the created session.

What do I have to do to get the node added on the server I specify with server="WAN" in the call to session.add_node?

Thanks for your help,

Andreas

```
#!/opt/core/venv/bin/python

from core.api.grpc import client
from core.api.grpc.wrappers import Position
from core.api.grpc.wrappers import Server

# interface helper
iface_helper = client.InterfaceHelper(ip4_prefix="10.0.0.0/24", ip6_prefix="2001::/64")

# create grpc client and connect
core = client.CoreGrpcClient()
core.connect()

# add session
session = core.create_session()

# set the servers 
session.servers =  [ Server(name="WAN", host="192.168.1.118") ]

# create nodes
position = Position(x=100, y=100)
# request node1 on server "WAN"
node1 = session.add_node(1, position=position, server="WAN")
position = Position(x=300, y=100)
node2 = session.add_node(2, position=position)

# create link
iface1 = iface_helper.create_iface(node1.id, 0)
iface2 = iface_helper.create_iface(node2.id, 0)
session.add_link(node1=node1, node2=node2, iface1=iface1, iface2=iface2)

# start session
core.start_session(session)

````

**Comment by bharnden** [comment, marked as answer]
That appears to be right, based on this example: https://github.com/coreemu/core/blob/develop/package/examples/grpc/distributed_switch.py

In theory that should work, I will have to get some time eventually to look into it.

Also for a node that is just an LXD/LXC or Docker container, you do not really need to use distributed and could just use a node of that type in the first place. There are a few requirements of the image for to help facilitate network orchestration.

**Reply by awolff64** [reply]
> That appears to be right, based on this example: https://github.com/coreemu/core/blob/develop/package/examples/grpc/distributed_switch.py
I should have looked at all the examples and my life would have been a bit easier ...

> In theory that should work, I will have to get some time eventually to look into it.
Thanks!

> Also for a node that is just an LXD/LXC or Docker container, you do not really need to use distributed and could just use a node of that type in the first place. There are a few requirements of the image for to help facilitate network orchestration.
I do not understand this. The nodes I set up are routers, and switches - so far no containers.
I run the daemon in a container because my host machine is ubuntu 22.10 and it allows me to easily test my distributed setup.
I tried remotes because I have a requirement for a distributed test topology that I want to control and configure from a central point.
